salt/transport/tcp.py

```diff
@@ -262,17 +373,94 @@ def _decode_messages(self, messages):
+                        try:
+                            byts = await self._stream.read_bytes(4096, partial=True)
+                        except tornado.iostream.StreamClosedError:
+                            log.trace("Stream closed, reconnecting.")
+                            stream = self._stream
+                            self._stream = None
+                            stream.close()
+                            if self.disconnect_callback:
+                                self.disconnect_callback()
+                            await self.connect()
+                            return
+                        self.unpacker.feed(byts)
+                        for msg in self.unpacker:
+                            framed_msg = salt.transport.frame.decode_embedded_strs(msg)
+                            return framed_msg["body"]
+        elif timeout:
+            try:
+                return await asyncio.wait_for(self.recv(), timeout=timeout)
+            except (
+                TimeoutError,
+                asyncio.exceptions.TimeoutError,
+                asyncio.exceptions.CancelledError,
+            ):
+                self.close()
+                await self.connect()
+                return
+        else:
+            for msg in self.unpacker:
+                framed_msg = salt.transport.frame.decode_embedded_strs(msg)
+                return framed_msg["body"]
+            while not self._closing:
+                async with self._read_in_progress:
+                    try:
+                        byts = await self._stream.read_bytes(4096, partial=True)
+                    except tornado.iostream.StreamClosedError:
+                        log.trace("Stream closed, reconnecting.")
+                        stream = self._stream
+                        self._stream = None
+                        stream.close()
+                        if self.disconnect_callback:
+                            self.disconnect_callback()
+                        await self.connect()
+                        log.debug("Re-connected - continue")
+                        continue
+                    self.unpacker.feed(byts)
+                    for msg in self.unpacker:
+                        framed_msg = salt.transport.frame.decode_embedded_strs(msg)
+                        return framed_msg["body"]
+
+    async def on_recv_handler(self, callback):
+        while not self._stream:
+            await asyncio.sleep(0.003)
+        while True:
+            # try:
+            msg = await self.recv()
+            if msg:
+                callback(msg)
 
     def on_recv(self, callback):
         """
-        Register an on_recv callback
+        Register a callback for received messages (that we didn't initiate)
         """
-        return self.message_client.on_recv(callback)
+        if self.on_recv_task:
+            self.on_recv_task.cancel()
+        if callback is None:
+            self.on_recv_task = None
```

Comment:
```suggestion
            self.on_recv_task = None
```
Assuming that `self.on_recv_task` was not previously `None` and that it was created using `asyncio.create_task` we should [`.cancel`](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel) it and we would still need to `await` for it ...

Reply:
I'm not sure we need to `await` canceled tasks.

Reply:
I'm thinking maybe we can add a task monitor class. We can add a periodic callback to run every once in a while and clean up tasks like this.

Reply:
> I'm not sure we need to `await` canceled tasks. If we do, it's going to cause a lot of problems. There are lots of places where we do not currently yield to tornado coroutines. As we migrate away from tornado issues like this are going to pop up and I was hoping we can leverage tasks for most of these cases.

If we can get away with it, ie, python let's us and it still works... I'm good.

Reply:
> I'm thinking maybe we can add a task monitor class. We can add a periodic callback to run every once in a while and clean up tasks like this.

Yeah, we could try this.